lib/memcached/elasticache.rb

```diff
@@ -0,0 +1,144 @@
+require 'memcached'
+require 'socket'
+require 'memcached/elasticache/version'
+require 'memcached/elasticache/auto_discovery/endpoint'
+require 'memcached/elasticache/auto_discovery/config_response'
+require 'memcached/elasticache/auto_discovery/stats_response'
+
+module Memcached
+  module Elasticache
+    class Client
+      attr_reader :endpoint, :options
+
+      def initialize(config_endpoint, options={})
+        @refresh_interval = options.delete(:refresh_interval) || 60
+        @max_retry_count = options.delete(:max_retry_count) || 1
+        @enable_legacy_ec = options.delete(:enable_legacy_ec) || false
+        @default_ttl = options[:ttl] || 0
+        @options = options
+
+        @last_updated_at = Time.now
+        @endpoint = Memcached::Elasticache::AutoDiscovery::Endpoint.new(config_endpoint, @enable_legacy_ec)
+        @client = Memcached::Client.new(cluster_servers, @options)
+      end
+
+      def clone
+        options = @options.dup.merge(
+          refresh_interval: @refresh_interval,
+          max_retry_count: @max_retry_count,
+          enable_legacy_ec: @enable_legacy_ec
+        )
+        Memcached::Elasticache::Client.new(config_endpoint, options)
+      end
+
+      # memcached client methods
+
+      def flush
+        refresh_with { @client.flush }
+      end
+
+      def set(key, value, ttl: @default_ttl, raw: false, flags: Memcached::Client::FLAGS)
+        refresh_with { @client.set(key, value, ttl: ttl, raw: raw, flags: flags) }
+      end
+
+      def get(key, raw: nil)
+        refresh_with { @client.get(key, raw: raw) }
+      end
+
+      def get_multi(keys, raw: nil)
+        refresh_with { @client.get_multi(keys, raw: raw) }
+      end
+
+      def delete(key)
+        refresh_with { @client.delete(key) }
+      end
+
+      def add(key, value, ttl: @default_ttl, raw: false, flags: Memcached::Client::FLAGS)
+        refresh_with { @client.add(key, value, ttl: ttl, raw: raw, flags: flags) }
+      end
+
+      def increment(key, offset = 1)
+        refresh_with { @client.increment(key, offset) }
+      end
+
+      def decrement(key, offset = 1)
+        refresh_with { @client.decrement(key, offset) }
+      end
+
+      def exist(key)
+        refresh_with { @client.exist(key) }
+      end
+
+      def replace(key, value, ttl: @default_ttl, raw: false, flags: Memcached::Client::FLAGS)
+        refresh_with { @client.replace(key, value, ttl: ttl, raw: raw, flags: flags) }
+      end
+
+      def prepend(key, value, ttl: @default_ttl, flags: Memcached::Client::FLAGS)
+        refresh_with { @client.prepend(key, value, ttl: ttl, flags: flags) }
+      end
+
+      def append(key, value, ttl: @default_ttl, flags: Memcached::Client::FLAGS)
+        refresh_with { @client.append(key, value, ttl: ttl, flags: flags) }
+      end
+
+      def namespace
+        @client.namespace
+      end
+
+      def namespace=(value)
+        @client.namespace = value
```

Comment:
`Memcached::Connection#set_prefix` finally calls `memcached_callback_set` in libmemcached.
[This method resets no connections](http://docs.libmemcached.org/memcached_callback.html), so we don't need to refresh server nodes.

However, namespace is stored in `Memcached::Client` object and it is destroyed after refresh.
I decide to remove `#namespace=` because we should set namespace in options.